hip/test/base/hip_executor.hip.cpp

```diff
@@ -0,0 +1,240 @@
+{
+    int count = 0;
+    hipGetDeviceCount(&count);
+    ASSERT_EQ(count, gko::HipExecutor::get_num_devices());
+}
+
+
+TEST_F(HipExecutor, AllocatesAndFreesMemory)
+{
+    int *ptr = nullptr;
+
+    ASSERT_NO_THROW(ptr = hip->alloc<int>(2));
+    ASSERT_NO_THROW(hip->free(ptr));
+}
+
+
+TEST_F(HipExecutor, FailsWhenOverallocating)
+{
+    const gko::size_type num_elems = 1ll << 50;  // 4PB of integers
+    int *ptr = nullptr;
+
+    ASSERT_THROW(
+        {
+            ptr = hip->alloc<int>(num_elems);
+            hip->synchronize();
+        },
+        gko::AllocationError);
+
+    hip->free(ptr);
+}
+
+
+__global__ void check_data(int *data)
+{
+    if (data[0] != 3 || data[1] != 8) {
+#if GINKGO_HIP_PLATFORM_HCC
+        asm("s_trap 0x02;");
+#else  // GINKGO_HIP_PLATFORM_NVCC
+        asm("trap;");
+#endif
+    }
+}
+
+TEST_F(HipExecutor, CopiesDataToHip)
+{
+    int orig[] = {3, 8};
+    auto *copy = hip->alloc<int>(2);
+
+    hip->copy_from(omp.get(), 2, orig, copy);
+
+    hipLaunchKernelGGL((check_data), dim3(1), dim3(1), 0, 0, copy);
+    ASSERT_NO_THROW(hip->synchronize());
+    hip->free(copy);
+}
+
+
+__global__ void init_data(int *data)
+{
+    data[0] = 3;
+    data[1] = 8;
+}
+
+TEST_F(HipExecutor, CopiesDataFromHip)
+{
+    int copy[2];
+    auto orig = hip->alloc<int>(2);
+    hipLaunchKernelGGL((init_data), dim3(1), dim3(1), 0, 0, orig);
+
+    omp->copy_from(hip.get(), 2, orig, copy);
+
+    EXPECT_EQ(3, copy[0]);
+    ASSERT_EQ(8, copy[1]);
+    hip->free(orig);
+}
+
+/* Properly checks if it works only when multiple GPUs exist */
+TEST_F(HipExecutor, PreservesDeviceSettings)
+{
+    auto previous_device = gko::HipExecutor::get_num_devices() - 1;
+    GKO_ASSERT_NO_HIP_ERRORS(hipSetDevice(previous_device));
+    auto orig = hip->alloc<int>(2);
+    int current_device;
+    GKO_ASSERT_NO_HIP_ERRORS(hipGetDevice(&current_device));
+    ASSERT_EQ(current_device, previous_device);
+
+    hip->free(orig);
+    GKO_ASSERT_NO_HIP_ERRORS(hipGetDevice(&current_device));
+    ASSERT_EQ(current_device, previous_device);
+}
+
+TEST_F(HipExecutor, RunsOnProperDevice)
+{
+    int value = -1;
+    GKO_ASSERT_NO_HIP_ERRORS(hipSetDevice(0));
+    hip2->run(ExampleOperation(value));
+    ASSERT_EQ(value, hip2->get_device_id());
+}
+
+TEST_F(HipExecutor, CopiesDataFromHipToHip)
+{
+    int copy[2];
+    auto orig = hip->alloc<int>(2);
+    GKO_ASSERT_NO_HIP_ERRORS(hipSetDevice(0));
+    hipLaunchKernelGGL((init_data), dim3(1), dim3(1), 0, 0, orig);
+
+    auto copy_hip2 = hip2->alloc<int>(2);
+    hip2->copy_from(hip.get(), 2, orig, copy_hip2);
+
+    // Check that the data is really on GPU2 and ensure we did not cheat
+    int value = -1;
+    GKO_ASSERT_NO_HIP_ERRORS(hipSetDevice(hip2->get_device_id()));
+    hipLaunchKernelGGL((check_data), dim3(1), dim3(1), 0, 0, copy_hip2);
+    GKO_ASSERT_NO_HIP_ERRORS(hipSetDevice(0));
+    hip2->run(ExampleOperation(value));
+    ASSERT_EQ(value, hip2->get_device_id());
```

Comment:
I sort of want to keep the logical structure of this code. In fact, both this big block and the next one are "assertions". The check is done in two steps, we check on the GPU side that we have the actual data from a kernel, we then check that it was run on the proper GPU by getting the device_id through `ExampleOperation` (a number of things could happen, by chance the data got there, we put it on the wrong GPU, ... even more so when considering a multi GPU scenario). Finally we copy the data back to CPU and check that it is correct there as a final step, which in itself ensures the GPU copies happen correctly. What I change according to the previous comment is I merge this code block with the other assertions at the bottom.

Reply:
That is fine with me. I did not consider the `GKO_ASSERT_NO_HIP_ERRORS` as test-asserts (since they are not google-test asserts), which is why I thought moving this assert to the others would group them nicely. However, it makes sense to consider this full block as one assert.